sql/core/src/main/scala/org/apache/spark/sql/internal/SharedState.scala

```diff
@@ -148,6 +149,8 @@ private[sql] class SharedState(val sparkContext: SparkContext) extends Logging {
 
 object SharedState {
 
+  URL.setURLStreamHandlerFactory(new SparkUrlStreamHandlerFactory())
```

Comment:
> Say one day we need to support a third-party file system named "shaofs"

So why not add the abstraction on that day? If `FsUrlStreamHandlerFactory` does what is needed here, and correctly support more than just the `hdfs` protocol (which isn't even enough to cover HDFS itself, which can be federated), then I don't see what the abstraction is buying us today.